contracts/cw-ics721-bridge/src/msg.rs

```diff
@@ -0,0 +1,189 @@
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::IbcTimeout;
+use cw721_proxy_derive::cw721_proxy;
+use cwd_interface::ModuleInstantiateInfo;
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    /// Code ID of cw721-ics contract. A new cw721-ics will be
+    /// instantiated for each new IBCd NFT classID.
+    ///
+    /// NOTE: this _must_ correspond to the cw721-base contract. Using
+    /// a regular cw721 may cause the ICS 721 interface implemented by
+    /// this contract to stop working, and IBCd away NFTs to be
+    /// unreturnable as cw721 does not have a mint method in the spec.
+    pub cw721_base_code_id: u64,
+    /// An optional proxy contract. If a proxy is set the contract
+    /// will only accept NFTs from that proxy. The proxy is expected
+    /// to implement the cw721 proxy interface defined in the
+    /// cw721-proxy crate.
+    pub proxy: Option<ModuleInstantiateInfo>,
+    /// Address that may pause the contract. PAUSER may pause the
+    /// contract a single time; in pausing the contract they burn the
+    /// right to do so again. A new pauser may be later nominated by
+    /// the CosmWasm level admin via a migration.
+    pub pauser: Option<String>,
+}
+
+#[cw721_proxy]
+#[cw_serde]
+pub enum ExecuteMsg {
+    /// Receives a NFT to be IBC transfered away. The `msg` field must
+    /// be a binary encoded `IbcAwayMsg`.
+    ReceiveNft(cw721::Cw721ReceiveMsg),
+
+    /// Pauses the bridge. Only the pauser may call this. In pausing
+    /// the contract, the pauser burns the right to do so again.
+    Pause {},
+
+    /// Mesages used internally by the contract. These may only be
+    /// called by the contract itself.
+    Callback(CallbackMsg),
+}
+
+#[cw_serde]
+pub enum CallbackMsg {
+    /// Mints a NFT of collection class_id for receiver with the
+    /// provided id and metadata. Only callable by this contract.
+    Mint {
+        /// The class_id to mint for. This must have previously been
+        /// created with `SaveClass`.
+        class_id: String,
+        /// Unique identifiers for the tokens.
+        token_ids: Vec<String>,
+        /// Urls pointing to metadata about the NFTs to mint. For
+        /// example, this may point to ERC721 metadata on IPFS. Must
+        /// be the same length as token_ids. token_uris[i] is the
+        /// metadata for token_ids[i].
+        token_uris: Vec<String>,
+        /// The address that ought to receive the NFTs. This is a
+        /// local address, not a bech32 public key.
+        receiver: String,
+    },
+    /// Much like mint, but will instantiate a new cw721 contract iff
+    /// the classID does not have one yet.
+    DoInstantiateAndMint {
```

Comment:
Why not just `InstantiateAndMint`? What's the significance of `Do..`?

Reply:
at the time, do seemed like good vibes. though now that you point it out, it's redundant. updated :)